src/input/input_python.rs

```diff
@@ -823,6 +823,14 @@ impl<'py> ValidatedDict<'py> for GenericPyMapping<'_, 'py> {
             Self::GetAttr(obj, _) => Ok(consumer.consume_iterator(iterate_attributes(obj)?)),
         }
     }
+
+    fn last_key(&self) -> Option<Self::Key<'_>> {
+        match self {
+            Self::Dict(dict) => dict.keys().iter().last(),
+            Self::Mapping(mapping) => mapping.keys().ok()?.iter().ok()?.last()?.ok(),
```

Comment:
NB both of these are potentially inefficient; `.keys()` creates a new `PyList` of all the keys.

Reply:
yup, I don't love it, but at least it's only called when `allow_partial=true`.

Is there a more efficient way? (especially for dicts?)

Long term, if we move to `RustModel` and iterating over dicts when building typeddicts, we should be able to minimise use of this.

Reply:
`.call_method0("keys").iter().last()` _might_ be better (Python keys iterator rather than list), but I think it still sucks. Can't get better until we iterate, as you say.

Reply:
I tested by adding:

```rs
#[pyfunction]
pub fn mapping_last_key_a<'a>(mapping: &'a Bound<'a, PyMapping>) -> Option<Bound<'a, PyAny>> {
    mapping.keys().ok()?.iter().ok()?.last()?.ok()
}

#[pyfunction]
pub fn mapping_last_key_b<'a>(mapping: &'a Bound<'a, PyMapping>) -> Option<Bound<'a, PyAny>> {
    mapping.call_method0(intern!(mapping.py(), "keys")).ok()?.iter().ok()?.last()?.ok()
}
```

and tested with

```py
from typing import Mapping
import timeit

from pydantic_core import _pydantic_core


class MyMapping(Mapping):
    def __init__(self, d):
        self._d = d

    def __getitem__(self, key):
        return self._d[key]

    def __iter__(self):
        return iter(self._d)

    def __len__(self):
        return len(self._d)


mapping = MyMapping({str(i): i for i in range(100)})

v = _pydantic_core.mapping_last_key_a(mapping)
assert v == '99', v
v = _pydantic_core.mapping_last_key_b(mapping)
assert v == '99', v


def run_bench(func):
    timer = timeit.Timer(
        "func(mapping)", setup="", globals={"func": func, "mapping": mapping}
    )
    n, t = timer.autorange()
    iter_time = t / n
    # print(f'{func.__module__}.{func.__name__}', iter_time)
    return int(iter_time * 1_000_000_000)


print(f'mapping_last_key_a: {run_bench(_pydantic_core.mapping_last_key_a)}ns')
print(f'mapping_last_key_b: {run_bench(_pydantic_core.mapping_last_key_b)}ns')
```

Output:

```
mapping_last_key_a: 2487ns
mapping_last_key_b: 1918ns
```

---

Outcome: they're both pretty slow, but your suggestion is slightly faster.